### PR TITLE
nixos/caddy: add support for Caddy v2

### PR DESCRIPTION
**Pull request by Br1ght0ne**
###### Motivation for this change
[Caddy v2](https://github.com/caddyserver/caddy/releases/tag/v2.0.0) was recently released.
While v2 support is necessary, configs for v1 shouldn't break (at least until 20.09).

###### Things done

This PR adds support for Caddy v2, and uses it by default, while still allowing users to opt-in to using Caddy v1.

Added `services.caddy.adapter` for supporting config formats other than Caddyfile.
See https://caddyserver.com/docs/config-adapters.

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by filalex77** [comment]
@GrahamcOfBorg test caddy

File: nixos/modules/services/web-servers/caddy.nix [hunk 1]
```diff
@@ -50,28 +79,37 @@ in {
...
+
+        Caddy v2 replaced CADDYPATH with XDG directories.
+        See https://caddyserver.com/docs/conventions#file-locations.
       '';
     };
 
     package = mkOption {
-      default = pkgs.caddy;
-      defaultText = "pkgs.caddy";
+      default = pkgs.caddy2;
```
Comment (xfix) [inline nixos/modules/services/web-servers/caddy.nix:89]: Hm, I'm wondering about this. Changing old configurations to `caddy2` is a breaking change, but at the same time, I'm well aware that Caddy probably won't be supported during 20.09 release lifetime, and probably will be removed.

I think defaulting to Caddy 2 is the right choice, but I would appreciate having the validation with `caddy validate` to inform about breakage during 20.03 to 20.09 upgrade.

Comment (filalex77) [reply nixos/modules/services/web-servers/caddy.nix:89]: @xfix Thanks, that's very helpful!

Could you please give me some guidance on how to plug `caddy validate` into all this, and properly report validation errors to the user? How can I check if, for instance, `runCommand` call was successful?

Comment (xfix) [reply nixos/modules/services/web-servers/caddy.nix:89]: To be fair, I don't exactly know how to use Nix, but you may want to take a look at `pkgs/build-support/writers/default.nix`.

I believe that any command that returns an error code will interrupt the build. `caddy validate` returns error code 1 on an error, so it's sufficient to run `caddy validate`.

Comment (filalex77) [reply nixos/modules/services/web-servers/caddy.nix:89]: @xfix So just failing the build is good enough? Okay, thanks!

File: nixos/modules/services/web-servers/caddy.nix [hunk 2]
```diff
@@ -5,6 +5,23 @@ with lib;
...
+  tlsConfig = {
+    apps.tls.automation.policies = [{
+      issuer = {
+        inherit (cfg) ca email;
+        module = "acme";
+      };
+    }];
+  };
+  adaptedConfig = importJSON (pkgs.runCommand "caddy-config-adapted.json" { } ''
+    ${cfg.package}/bin/caddy adapt --validate \
```
Comment (xfix) [inline nixos/modules/services/web-servers/caddy.nix:20]: Actually, thinking about it, I think `caddy adapt` should be sufficient for a syntax check, it shouldn't be additionally validated. The reason being is that validation check may complain about files not existing inside build environment.

**Comment by filalex77** [comment]
@xfix I removed the commit adding `--validate` and rebased onto latest `master`.

File: nixos/modules/services/web-servers/caddy.nix [hunk 3]
```diff
@@ -5,6 +5,24 @@ with lib;
...
+        inherit (cfg) ca email;
+        module = "acme";
+      };
+    }];
+  };
+  adaptedConfig = importJSON (pkgs.runCommand "caddy-config-adapted.json" { } ''
+    ${cfg.package}/bin/caddy adapt \
+      --config ${configFile} --adapter ${cfg.adapter} > $out
+  '');
+  # TODO: validate with `caddy validate`?
```
Comment (xfix) [inline nixos/modules/services/web-servers/caddy.nix:23]:
```suggestion
```

**Review by sugar700 — CHANGES_REQUESTED** [review]
This looks fine, it's mergeable after the TODO comment gets deleted.

File: nixos/modules/services/web-servers/caddy.nix (hunk 3 above)
Comment (curbengh) [reply nixos/modules/services/web-servers/caddy.nix:23]: to add on, `adapt` validates the input config by default and will return error 1 when encounter invalid input.

**Comment by AluisioASG** [comment]
`ExecReload` has to be [updated](https://caddyserver.com/docs/v2-upgrade#command-line) as well:

> Signals are mostly the same, except USR1 and USR2 are no longer supported. Use the `caddy reload` command or the API instead to load new configuration.

**Comment by k2s** [comment]
There are 2 official upstream systemd service files with explanation at https://github.com/caddyserver/dist/tree/master/init.

In compare to v1 the Caddy v2 starts without configuration file and best practice in v2 is to use API to upload config. JSON is native format, but with simple curl POST it is possible to upload regular Caddyfile too. Caddy reloads last state after restart. All that is working with https://github.com/caddyserver/dist/blob/master/init/caddy-api.service.

In my opinion that this should be default in NixOS. The configuration file should be added to command line arguments only if additional parameters related to configuration source and provider are specified.

**Comment by sephii** [comment]
It would be really nice to have this in 20.09. :) can I help in some way?

**Comment by xfix** [comment]
@filalex77 What is the status of this pull request. I really would like to have this in NixOS 20.09.

**Review by sugar700 — CHANGES_REQUESTED** [review]
`ExecReload = "${pkgs.coreutils}/bin/kill -USR1 $MAINPID";` should be changed to use `caddy reload --config ${path}` instead for Caddy 2 as USR1 signal is no longer supported.

**Comment by sephii** [comment]
I've opened a new PR which addresses the remaining comments in the hope of getting this in 20.09. I've also rebased on current master. I still have problems with running tests though, any help is appreciated.

**Comment by filalex77** [comment]
Closed in favor of #97217 by @sephii.

**Pull request closed without merging**